modules/standard/IO.chpl

```diff
@@ -5229,6 +5229,149 @@ proc _channel.writeBinary(arg:numeric, endian:ioendian) throws {
   }
 }
 
+/*
+   Write a :type:`~String.string` to a fileWriter in binary format
+
+   :arg s: the ``string`` to write
+   :arg size: the number of codepoints to write from the ``string``
+
+   :throws SystemError: Thrown if the string could not be written to the fileWriter.
+   :throws IllegalArgumentError: Thrown if ``size`` is larger than ``s.size``
+*/
+proc fileWriter.writeBinary(s: string, size: int = s.size) throws {
+  // handle bad arguments
+  if size > s.size then
+    throw new owned IllegalArgumentError("size", "cannot exceed length of provided string");
+  if s.hasEscapes then
+    throw createSystemError(EILSEQ, "illegal use of escaped string characters in 'writeBinary'");
+
+  // count the number of bytes to write
+  var sLocal = s.localize();
+  var bytesLen = 0;
+  if size == sLocal.size {
+    bytesLen = s.numBytes;
+  } else {
+    for ((_, cp_byte_len), i) in zip(sLocal._indexLen(), 0..) {
+      if i == size then break;
+      bytesLen += cp_byte_len;
+    }
+  }
+
+  // write the first bytesLen bytes of the string to the fileWriter
+  var e: errorCode = qio_channel_write_string(
+    false,
+    iokind.native: c_int,
+    qio_channel_str_style(this._channel_internal),
+    this._channel_internal,
```

Comment:
The `_channel_internal` field is a C pointer. If you call this `writeBinary` from a different locale from the one where the channel was created, the `_channel_internal` field will point to memory on another locale, and this whole thing will core dump. So I think you need *both* the `on` statement as well as the `s.localize()`. (The `s.localize()` part is for getting the string data on to the same locale as the channel, so the string data pointer can be passed to the C routine without core dump).